build: Add -Werror=implicit-fallthrough compile flag

hebasto opened the pull request.

hebasto committed:
Use C++17 [[fallthrough]] attribute, and drop -Wno-implicit-fallthrough

hebasto committed:
build: Add -Werror=implicit-fallthrough compile flag

practicalswift commented:
cr ACK e95d14e07e9215d6d8ba1a56d13485a18e3550ba: patch looks correct!

I've always been worried about introduction of bugs in our code due to unintentional fallthroughs. Thanks for reducing the number of things to worry about by one :)

MarcoFalke commented on configure.ac:
```diff
@@ -439,6 +439,7 @@ if test "x$enable_werror" = "xyes"; then
+  AX_CHECK_COMPILE_FLAG([-Werror=implicit-fallthrough], [ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=implicit-fallthrough"], [], [$CXXFLAG_WERROR])
 fi
 
 if test "x$CXXFLAGS_overridden" = "xno"; then
```
any reason to not enable the warning as well?

hebasto replied:
You are right!

From https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html:
> `-Wimplicit-fallthrough` is the same as `-Wimplicit-fallthrough=3`...
>
> The `-Wimplicit-fallthrough=3` warning is enabled by `-Wextra`.

But that is _not_ the case for clang. Going to fix it now.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21430#issuecomment-799583722).

maflcko reviewed (approved):
cr ACK e95d14e07e9215d6d8ba1a56d13485a18e3550ba

didn't review the leveldb build change

hebasto commented:
Updated e95d14e07e9215d6d8ba1a56d13485a18e3550ba -> 2f3436bcad878a313ce53bab2fc8a2fed9056351 ([pr21430.01](https://github.com/hebasto/bitcoin/commits/pr21430.01) -> [pr21430.02](https://github.com/hebasto/bitcoin/commits/pr21430.02), [diff](https://github.com/hebasto/bitcoin/compare/pr21430.01..pr21430.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/21430#discussion_r594495759):
> any reason to not enable the warning as well?

@MarcoFalke 
> didn't review the leveldb build change

https://github.com/bitcoin/bitcoin/blob/c771fc0dc1bcb48fc6aa77b61c1ff31742ac8bb7/src/leveldb/util/hash.cc#L11-L12

practicalswift commented:
cr re-ACK 2f3436bcad878a313ce53bab2fc8a2fed9056351

Consider submitting the `tinyformat.h` changes upstreams too (https://github.com/c42f/tinyformat).

hebasto commented:
Rebased 2f3436bcad878a313ce53bab2fc8a2fed9056351 -> 5bc16d28e73a068382e66fce9ffcb604a3db34dc ([pr21430.02](https://github.com/hebasto/bitcoin/commits/pr21430.02) -> [pr21430.03](https://github.com/hebasto/bitcoin/commits/pr21430.03)) due to the conflicts with #21610 and #21613.

practicalswift commented:
cr ACK 5bc16d28e73a068382e66fce9ffcb604a3db34dc: patch looks correct

Thanks a lot for doing this @hebasto: the introduction of `-Werror=implicit-fallthrough` in our project combined with proper `[[fallthrough]]` annotations will most likely kill the  "unintended fallthroughs" bug class for good.

Killing bug instances is cool, but killing entire bug classes is even cooler! :)

hebasto commented:
Rebased 5bc16d28e73a068382e66fce9ffcb604a3db34dc -> 45dc9f383cf46a556fefa609dc88135cfe55b5bc ([pr21430.03](https://github.com/hebasto/bitcoin/commits/pr21430.03) -> [pr21430.04](https://github.com/hebasto/bitcoin/commits/pr21430.04)) due to the conflict with #22258.

practicalswift commented:
cr re-ACK 45dc9f383cf46a556fefa609dc88135cfe55b5bc

fanquake commented on configure.ac:
```diff
@@ -1245,7 +1246,7 @@ AC_LINK_IFELSE(
 AC_DEFINE([HAVE_SYSTEM], [HAVE_STD__SYSTEM || HAVE_WSYSTEM], [std::system or ::wsystem])
 
-LEVELDB_CPPFLAGS=
+LEVELDB_CPPFLAGS="-DFALLTHROUGH_INTENDED=@<:@@<:@fallthrough@:>@@:>@"
```
Shouldn't this be done in `Makefile.leveldb.include`?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21430#issuecomment-873435083).

fanquake reviewed (commented):
Concept ACK.

hebasto commented:
Updated 45dc9f383cf46a556fefa609dc88135cfe55b5bc -> ebc36d94cbc927a5c68873b92442937c420f317f ([pr21430.04](https://github.com/hebasto/bitcoin/commits/pr21430.04) -> [pr21430.05](https://github.com/hebasto/bitcoin/commits/pr21430.05)):

- rebased due to the conflict with #19438
- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/21430#discussion_r661261506)

fanquake commented on src/Makefile.leveldb.include:
```diff
@@ -11,6 +11,7 @@ EXTRA_LIBRARIES += $(LIBMEMENV_INT)
 LIBLEVELDB += $(LIBLEVELDB_INT) $(LIBCRC32C)
 LIBMEMENV += $(LIBMEMENV_INT)
 
+LEVELDB_CPPFLAGS += -DFALLTHROUGH_INTENDED=[[fallthrough]]
```
This should be in `LEVELDB_CPPFLAGS_INT` and down with the other defines.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21430#issuecomment-873826524).

hebasto commented:
Updated ebc36d94cbc927a5c68873b92442937c420f317f -> 3c4c8e79baf02af97ba1502189f649b04ef2198d ([pr21430.05](https://github.com/hebasto/bitcoin/commits/pr21430.05) -> [pr21430.06](https://github.com/hebasto/bitcoin/commits/pr21430.06), [diff](https://github.com/hebasto/bitcoin/compare/pr21430.05..pr21430.06)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/21430#discussion_r663583387)

fanquake reviewed (approved):
ACK 3c4c8e79baf02af97ba1502189f649b04ef2198d - looks ok to me now. Checked that warnings occur in our code & leveldb by removing a `[[fallthrough]]` or `FALLTHROUGH_INTENDED`.

theStack commented:
Concept ACK

jarolrod reviewed (commented):
ACK 3c4c8e79baf02af97ba1502189f649b04ef2198d

This is nice to be able to detect fallthroughs going forward. Appropriate since GCC 7 is the current minimum supported GCC. 

Tested that the codebase itself will throw the warning by following the same step of removing one of the added `[[fallthrough]]` statements as outlined here: https://github.com/bitcoin/bitcoin/pull/21430#pullrequestreview-702673309

hebasto commented:
@practicalswift @MarcoFalke 

Do you mind making another (final?) review?

theStack reviewed (approved):
ACK 3c4c8e79baf02af97ba1502189f649b04ef2198d

Tested that removing an arbitrary `[[fallthrough]]` statment in the codebase leads to either a warning (default configuration) or an error (if configured with `--enable-werror`) with clang 11.0.1-2 (Debian bullseye/sid), e.g.:
```
hash.cpp:52:9: error: unannotated fall-through between switch labels [-Werror,-Wimplicit-fallthrough]
        case 2:
        ^
hash.cpp:52:9: note: insert '[[fallthrough]];' to silence this warning
        case 2:
        ^
        [[fallthrough]];
hash.cpp:52:9: note: insert 'break;' to avoid fall-through
        case 2:
        ^
        break;
1 error generated.
```

By the way, any reason we don't `--enable-werror` by default? Being on the cautious side would seem to be a more sane default to me.

MarcoFalke commented:
> By the way, any reason we don't --enable-werror by default? Being on the cautious side would seem to be a more sane default to me.

There might be unknown warnings in future releases of compilers that we are not aware of. Also, compilers might produce false warnings if they have bugs. So we only error on well tested and selected warnings.
